### ISSUE TITLE
"Updating..." not shown when "refresh=true" items are clicked

### ISSUE THREAD

**Issue by seanw** [matryer/xbar#622]
## Versions

Xbar version: 2.0.15-beta
MacOS version: 11.2.1

## Setup 

Create this test plugin script:

```
#!/bin/bash

sleep 1 # to simulate a slow operation like a network request

echo `date`
echo "---"
echo "Refresh | refresh=true"
```

## Steps

1. Refresh xbar and the test plugin will show the current time in xbar.

2. Click the plugin menu and select "Refresh".

## What you'll see

The displayed time freezes for 1 second and then updates.

## What I expected to see

The time to be replaced with "Updating..." then after 1 second this would be replaced with the current date time.

This is the old bitbar behaviour and in my opinion gives much better feedback that an update is in progress. Right now, you might think the item you clicked on didn't work, you clicked the wrong item or xbar froze/crashed. This is much worse when the refresh operation takes longer e.g. 10 seconds for a network request.

A feature suggestion would be to add a way to change the default "Updating..." message and perhaps a new plugin parameter that lets you customise the update message per item to something with more context e.g. `Refresh | refresh=true refreshMessage="Fetching time..."`.

## Extra

Congrats on the relaunch and thanks for the awesome software! 

By the way, I would set a GitHub issue template because I wasn't sure what information to give. Hope the above is simple to follow.

**Comment by matryer** [comment]
@seanw thanks for the excellent report, I'll look at this today.

**Comment by magicus** [comment]
As a side note, I was always annoyed by the long "Updating ..." message in BitBar that shuffled around icons in the tray quite a lot. :-(

I agree with the reporter that an update being performed should show up in the UI, but perhaps this can be shown in a more compact way? Ideal would be some kind of "spinning wheel" throbber, it is compact, and the animation shows that something is in progress. (But I can also see how this is harder to implement than just outputting some text.) Or, as a compromise, just replace it with `...`?

**Comment by matryer** [comment]
@magicus agree re the "Refreshing..." text - it was annoying, I will work on a new solution to this.

The goal is to show that an update is happening, without annoying the user. This will only happen if the user selects the refresh. For passive refreshing, it will silently swap the menus rather than indicate the pending state.

Sound good?

**Comment by magicus** [comment]
Agree, sounds good. For background updates, the old state remains until a new can be presented. (That's kind of the core idea of xbar with periodic updates...)

**Comment by matryer** [comment]
@seanw Thanks again for reporting. Can you confirm this is better in the latest release and we can close this issue?

**Comment by seanw** [comment]
@matryer

1. I'm on version v2.0.19-beta using the same script, and not seeing the "..." when I click the "Refresh | refresh=true" menu item. It just freezes then updates like before.

2. However, if I click the system tray and press "CMD+R", or select the "Preferences > Refresh" menu item, I do see the "...".  But this behaviour needs to happen for the "refresh=true" item above as well?

Just throwing ideas out there, but maybe dimming the opacity of the system tray text while it's updating might be worth exploring if you don't like the width suddenly changing? Not sure if it'll be obvious enough though and it might be a problem for already dim text and short text. The script dev could simulate this themselves with the `refreshText` customisation idea + ANSI colours.

**Comment by matryer** [comment]
@seanw You were right, we weren't using the same refresh path if the refresh was triggered via `refresh=true` option.

That's fixed now, I wonder if you could verify for me? [xbar.v2.0.19-beta-5-g9a866463.tar.gz](https://github.com/matryer/xbar/files/6117304/xbar.v2.0.19-beta-5-g9a866463.tar.gz)

**Comment by seanw** [comment]
@matryer 

- Nice, xbar.v2.0.19-beta-5-g9a866463.tar.gz fixes the issue in the original report. 😄

- Not sure if it's related to this one https://github.com/matryer/xbar/issues/620 but I tried this: 
```
#!/bin/bash

sleep 2

echo `date`
echo "---"
echo 'ItemInstant | refresh=true'
echo "ItemShell | bash='$0' refresh=true terminal=false"
```

`ItemInstant` shows `...` as expected. I expected `ItemShell` to behave the same but it doesn't appear to do anything - the time on the system tray never updates and I don't see `...`.

**Comment by seanw** [comment]
@matryer I was expecting:

1. User clicks `ItemShell`

2. The system tray shows `...` and the `shell` param is run.

3. Once the `shell` param is complete, the plugin refreshes and the system tray gets updated (in this case, showing the up-to-date time).

It does something like this with the original bitbar but to be honest I'm not sure if using `$0` for the `bash` param is making the plugin script run once for `bash` and then again for the refresh. 

Anyway, happy to open another issue with a test case if this is getting out of scope and the above is actually an issue.

**Comment by matryer** [comment]
@seanw I've opened https://github.com/matryer/xbar/issues/626. Meanwhile, can you confirm that this issue is fixed now? We opted for `...` but it is clear now that the plugin is refreshing.

**Comment by seanw** [comment]
@matryer Fixed. Thanks for your quick resolution! Happy to test #626 too if you want to ping me later if that's helpful.
